backend/fms_core/viewsets/full_sample.py

```diff
@@ -0,0 +1,216 @@
+from collections import Counter
+from django.core import serializers
+import json
+
+from rest_framework import viewsets
+from rest_framework.decorators import action
+from rest_framework.response import Response
+from django.db.models import Count, Q, F
+from django.core.exceptions import ValidationError
+
+from datetime import datetime
+from ..utils import RE_SEPARATOR, float_to_decimal
+
+
+from fms_core.models import FullSample, Biosample, Sample, DerivedSample, SampleKind, Container, Individual, DerivedBySample
+from fms_core.serializers import FullSampleSerializer, FullSampleExportSerializer, SampleSerializer
+from fms_core.template_importer.importers import SampleSubmissionImporter, SampleUpdateImporter
+
+from fms_core.template_paths import SAMPLE_SUBMISSION_TEMPLATE, SAMPLE_UPDATE_TEMPLATE
+
+from ._constants import _full_sample_filterset_fields
+from ._utils import TemplateActionsMixin, _list_keys, versions_detail
+from fms_core.services.sample import create_full_sample
+
+class FullSampleViewSet(viewsets.ModelViewSet, TemplateActionsMixin):
+    queryset = FullSample.objects.select_related("individual", "container", "sample_kind", "biosample", "derived_sample").all()
+    serializer_class = FullSampleSerializer
+
+    ordering_fields = (
+        *_list_keys(_full_sample_filterset_fields),
+    )
+
+    filterset_fields = {
+        **_full_sample_filterset_fields,
+    }
+
+    template_action_list = [
+        {
+            "name": "Add Samples",
+            "description": "Upload the provided template with up to 384 new samples.",
+            "template": SAMPLE_SUBMISSION_TEMPLATE,
+            "importer": SampleSubmissionImporter,
+        },
+        {
+            "name": "Update Samples",
+            "description": "Upload the provided template with up to 384 samples to update.",
+            "template": SAMPLE_UPDATE_TEMPLATE,
+            "importer": SampleUpdateImporter,
+        }
+    ]
+
+    def create(self, request, *args, **kwargs):
+        error = {}
+        full_sample = request.data
+
+        #Container related information
+        try:
+            container_obj = Container.objects.get(pk=full_sample['container'])
```

Comment:
I would not manually test each error ... you are duplicating the "clean" behaviour. i would create the biosample, derived sample and sample. The clean used to return the correct error mapping you can keep using it. It also creates some overhead (not much but if it is not necessary...)

Reply:
What do you mean by each error? Container, Individual and Sample Kind are given directly by the frontend (id's) so I need to fetch the objects anyways to be able to create biosample/derived/sample. Or you mean not to wrap them in a try/except block?

Reply:
you receive the primary key from the frontend for each of container individual and sample kind ... you do not need to get them. you just need to set new_sample.individual_id = full_sample["individual"]. If the given ID is not correct for some reason you'll get an "individual" error back from the model that you can catch and forward to the front end.